futher scheduler changes for invoke_quant: prologue low prec, (slightly) more aggressive fusion

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145104

Respect invoke_quant low precision options, also, be more aggressive in attepmting fusion.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

shunting314 commented on torch/_inductor/scheduler.py:
```diff
@@ -3477,36 +3521,7 @@ def can_fuse(self, node1: BaseSchedulerNode, node2: BaseSchedulerNode) -> bool:
                 )
```
Maybe move the whole if block regarding prologue fusion to `can_fuse_prologue` to make `can_fuse` smaller.

shunting314 commented on torch/_inductor/scheduler.py:
```diff
@@ -3384,6 +3384,50 @@ def unfusable_node(self, node: BaseSchedulerNode) -> bool:
+        """
+        Heuristics to avoid benchmarking predictably slow prologue fusions
+        """
+        # user opt into more aggressive prologue fusion, dont use heuristics
```
User opt in by using `invoke_quant`?

jansel reviewed (changes requested):
Failing tests?

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

jansel commented on torch/_inductor/graph.py:
```diff
@@ -492,6 +492,10 @@ def __init__(
         self.aligned_inputs = OrderedSet[str]()
         self.no_fuse_buffer_names = OrderedSet[str]()
 
+        self.low_precision_codegen_ops = OrderedSet[str]()
+        # more aggressive prologue fusion
+        self.invoke_quant_ops = OrderedSet[str]()
```
```suggestion
        self.low_precision_codegen_ops : OrderedSet[str] = OrderedSet()
        # more aggressive prologue fusion
        self.invoke_quant_ops : OrderedSet[str] = OrderedSet()
```
to avoid runtime calls to `__getitem__`.

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update on "futher scheduler changes for invoke_quant: prologue low prec, (slightly) more aggressive fusion"


Respect invoke_quant low precision options, also, be more aggressive in attepmting fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

eellison committed:
Update on "futher scheduler changes for invoke_quant: prologue low prec, (slightly) more aggressive fusion"


Respect invoke_quant low precision options, also, be more aggressive in attepmting fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

eellison committed:
Update on "futher scheduler changes for invoke_quant: prologue low prec, (slightly) more aggressive fusion"


Respect invoke_quant low precision options, also, be more aggressive in attepmting fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
